Ensure @params in shadow resource always has a valid value.

miah opened the pull request:
Add tests for method chained shadow resource with readable and unreadable shadow files.

Ensure @params always has a safe value, otherwise we may stacktrace when
unable to read /etc/shadow and invoked with method chaining.

My only complaint with this is that when the `/etc/shadow` file is unreadable, you won't see the error when chaining; you'll need to test the file or a specific user.

<img width="1450" alt="image" src="https://user-images.githubusercontent.com/49856/38575182-aa1617ca-3caf-11e8-85f2-466d93832890.png">

Signed-off-by: Miah Johnson <miah@chia-pet.org>

Fixes #2913

miah committed:
Add tests for method chained shadow resource with readable and
unreadable shadow files.

Ensure @params always has a safe value, otherwise we may stacktrace when
unable to read /etc/shadow and invoked with method chaining.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

jerryaldrichiii reviewed (approved):
Looks good to me! Thanks @miah!

TrevorBramble commented on lib/resources/shadow.rb:
```diff
@@ -43,7 +42,7 @@ def initialize(path = '/etc/shadow', opts = nil)
       @raw_content = opts[:content] || read_file_content(@path, allow_empty: true)
       @lines = @raw_content.to_s.split("\n")
-      @params = @lines.map { |l| parse_shadow_line(l) }
+      params
```
I'd really rather see this assignment preserved, rather than treating the setup as a side-effect of calling a method.

```ruby
def initialize
  @params = set_params(@lines)
end

def set_params(lines)
  lines.nil? ? [] : lines.map { |l| parse_shadow_line(1) }
end
```

TrevorBramble replied:
While we're there, `@raw_content` and `@lines` are only used in the initializer (if you pass `@lines` to the `params` method) so why not make them local variables instead? You could even inline the lines parsing because there's no conditional logic involved.
```
def initialize(path = '/etc/shadow', opts = nil)
  # ...
  params(raw_content.to_s.split("\n"))
end
```

...And while we're there
```ruby
def initialize(path = '/etc/shadow', opts = nil)
  opts ||= {}
  # ...
end
```
This makes me so sad. Would you mind?
```ruby
def initialize(path = '/etc/shadow', opts = {})
  # ...
end
```

miah replied:
I like this `@params = params(@raw_content.to_s.split("\n"))` but when I set up our params this way, when `filter` is called I'm still getting a nil value.

```
NoMethodError: undefined method `find_all' for nil:NilClass
  /Users/miah/projects/github/tmp/inspec/lib/resources/shadow.rb:80:in `block in filter'
```

So what I've done is preserved `@lines` and reference that from set_params.

TrevorBramble replied:
Ah, I missed `def lines` elsewhere. Still, I'd pass it into the new method that makes your params and use it as a local variable there.

TrevorBramble commented on lib/resources/shadow.rb:
```diff
@@ -121,10 +120,14 @@ def to_s
 
+    def params
+      @params = @lines.nil? ? [] : @lines.map { |l| parse_shadow_line(l) }
+    end
```
Couple of directions this could go to make the point more obvious.

You could separate the guard statement:
```ruby
def params(lines)
  return [] if lines.nil?

  lines.map { |l| parse_shadow_line(1) }
end
```
And that maps directly to your logic in the ternary. But what I think you're actually concerned about is calling `map` on `nil`? So this would actually be more direct and clear:
```ruby
def params(lines)
  Array(lines).map { |l| parse_shadow_line(1) }
end
```
`Array(nil)` will produce the empty array, and pass through the `map` call to return what you expect, just that empty array. So there's one path through the code and all the needed safety.

TrevorBramble commented on lib/resources/shadow.rb:
```diff
@@ -121,10 +120,14 @@ def to_s
 
     def map_data(id)
-      @params.map { |x| x[id] }
+      params.map { |x| x[id] }
```
Consider using `collect` here instead of `map` to make it clear that you're collecting a new set of values from the input data. (Same for the parsed shadow lines in the `params` method.)

TrevorBramble commented on test/unit/resources/shadow_test.rb:
```diff
@@ -86,6 +86,24 @@
 
+  describe 'when method chained' do
+    let(:unreadable_shadow) { load_resource('shadow', '/etc/unreadable_shadow') }
+    it 'can read /etc/shadow and #filter matches user with no password and inactive_days' do
```
Separate sections with newlines please. =^)

TrevorBramble reviewed (changes requested):
You asked for it! ;^)

miah committed:
Wrap deprecation notices with a proc/must_output to clean up test
output.

Added some missing newlines.

Catch deprecation notice on `lines`.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Resolve the majority of the issues pointed out by @tbramble.

Deprecate `lines`; its really only used internally but it was 'exposed'
through tests and who knows if there is external use. `lines` is not
documented as a property at least..

`#set_params` is much better now =)

Signed-off-by: Miah Johnson <miah@chia-pet.org>

TrevorBramble commented on lib/resources/shadow.rb:
```diff
@@ -35,15 +35,13 @@ class Shadow < Inspec.resource(1)
 
-    def initialize(path = '/etc/shadow', opts = nil)
-      opts ||= {}
+    def initialize(path = '/etc/shadow', opts = {})
```
Thank you!

chris-rock commented on lib/resources/shadow.rb:
```diff
@@ -35,15 +35,13 @@ class Shadow < Inspec.resource(1)
-    def initialize(path = '/etc/shadow', opts = nil)
-      opts ||= {}
+    def initialize(path = '/etc/shadow', opts = {})
       @path = path || '/etc/shadow'
```
I am wondering if we could check if the file exist and throw an exception here? Otherwise we absorb the error.

miah replied:
Definitely can do this, but the file resource we use for reading and content gathering also does this; but we do absorb the error when method chained.

chris-rock commented on lib/resources/shadow.rb:
```diff
@@ -35,15 +35,13 @@ class Shadow < Inspec.resource(1)
-      @raw_content = opts[:content] || read_file_content(@path, allow_empty: true)
-      @lines = @raw_content.to_s.split("\n")
-      @params = @lines.map { |l| parse_shadow_line(l) }
+      raw_content = opts[:content] || read_file_content(@path, allow_empty: true)
```
I think it would be beneficial to move the read logic out of the initialize function and do this during runtime not object creation time. This would allow us to do less during `inspec check` runs.

miah committed:
Move params setup entirely out of initialize.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Use entries rather than params.
Update test output to match updated entries.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

TrevorBramble reviewed (approved):
<emoji>

jquick reviewed (approved):
Thanks @miah
